### PR TITLE
fetchgit: add tag argument

### PR DESCRIPTION
**Pull request by jtojnar**
###### Motivation for this change
While it is possible to pass `rev = tag` to `fetchgit`, it may result in ambiguity when here is a branch with the same name as the tag. To be sure, we need to use `rev = "refs/tags/${tag}"`.

Since that is quite ugly, this commit introduces a shortcut, allowing us to just `inherit tag`.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/build-support/fetchgit/default.nix [hunk 1]
```diff
@@ -45,11 +45,17 @@ in
 
 assert deepClone -> leaveDotGit;
 
-if md5 != "" then
+if rev != null && tag != null then
+  throw "fetchgit cannot take both tag and name arguments"
```
Comment (timokau) [inline pkgs/build-support/fetchgit/default.nix]: `name` is also mutually exclusive with `tag` and `rev` right?

**Review by timokau — COMMENTED** [review]
Good idea! This should probably be documented somewhere right? Will this also indirectly affect `fetchFromGitHub`?

File: pkgs/build-support/fetchgit/default.nix (hunk 1 above)
Comment (jtojnar) [reply pkgs/build-support/fetchgit/default.nix]: Sorry, that was a mistake. `name` is independent of these arguments.

**Comment by timokau** [comment]
Well thats one upside of not having documentation I guess. Nothing to update.

Should we then do the same thing with `fetchFrom...`? Would be good for consistency and I feel like those are used more often than `fetchgit` (although a grep through the repository tells me otherwise).

**Comment by jtojnar** [comment]
Added the argument to `fetchFromGit{Hub,Lab}`.

Needed to escape the `rev` in `fetchFromGitLab` to avoid errors like:

building '/nix/store/cxfkh2lgfbhhpgcphsh8wkn2j25y50hn-source.drv'...

```
trying https://gitlab.gnome.org/api/v4/projects/gnumdk%2Flollypop-portal/repository/archive.tar.gz?sha=refs/tags/0.9.7
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 15537  100 15537    0     0  15011      0  0:00:01  0:00:01 --:--:-- 15011
unpacking source archive /build/0.9.7
do not know how to unpack source archive /build/0.9.7
builder for '/nix/store/cxfkh2lgfbhhpgcphsh8wkn2j25y50hn-source.drv' failed with exit code 1
```

Also I am not sure about what should the `rev` introduced with https://github.com/NixOS/nixpkgs/pull/9273 and https://github.com/NixOS/nixpkgs/pull/15139 be. I used `refs/tags/${tag}` for now.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -283,10 +289,10 @@ with pkgs;
       netrcImpureEnvVars = [ "${varBase}USERNAME" "${varBase}PASSWORD" ];
     };
     fetcherArgs = (if fetchSubmodules
-        then { inherit rev fetchSubmodules; url = "${baseUrl}.git"; }
-        else ({ url = "${baseUrl}/archive/${rev}.tar.gz"; } // privateAttrs)
+        then { inherit rev tag fetchSubmodules; url = "${baseUrl}.git"; }
+        else ({ url = "${baseUrl}/archive/${revOrTag}.tar.gz"; } // privateAttrs)
```
Comment (timokau) [inline pkgs/top-level/all-packages.nix]: Did you forget to escape here?

Comment (jtojnar) [reply pkgs/top-level/all-packages.nix]: GitHub does not seem to need escaping. GitLab does not need it either but fetchzip does, in order to be able to extract the file.

Comment (timokau) [reply pkgs/top-level/all-packages.nix]: Oh nevermind I see the difference. Isn't there still some escaping necessary, just not of the `/`?

**Review by timokau — APPROVED** [review]

**Comment by marsam** [comment]
> While it is possible to pass rev = tag to fetchgit, it may result in ambiguity when here is a branch with the same name as the tag. To be sure, we need to use rev = "refs/tags/${tag}".

AFAIK tags have precedence over a branch refs https://git-scm.com/docs/git-rev-parse.html#_specifying_revisions, i.e. `rev = "refs/tags/${tag}"` should not be necessary

**Comment by jtojnar** [comment]
Then I do not understand why `refs/tags/${version}` is so frequent in nixpkgs.

File: lib/strings.nix [hunk 3]
```diff
@@ -315,6 +315,23 @@ rec {
...
+
+     https://en.wikipedia.org/wiki/Percent-encoding
+
+     Type: string -> string
+
+     Example:
+       escapeUriSegment "foo/bar?baz=qux&quux=quuz#corge"
+       => "foo%2Fbar%3Fbaz%3Dqux%26quux%3Dquuz%23corge"
+  */
+  escapeUriSegment = arg:
```
Comment (edolstra) [inline lib/strings.nix]: This sort of thing should not be done in Nix code because it's slow. You can also do escaping in the builder.

**Comment by edolstra** [comment]
Not really in favor of this because it makes `fetchgit` even more complex, and adds another inconsistency with `builtins.fetchGit`.

**Pull request closed without merging**